### ISSUE TITLE
php70u-pecl-amqp

### ISSUE THREAD

**Issue by marksteele** [iuscommunity/wishlist#52]
### What package do you want?

php70u-pecl-amqp 1.6.1 (or whichever version works)

### Why?

Not currently available.

### OS versions

RHEL/CentOS 7

### Testing

I agree to test the new package to ensure that it works as expected.

**Comment by carlwgeorge** [comment]
Thanks for this request @marksteele.  I'll get to work on new packages.

**Comment by marksteele** [comment]
Following specfile spits out an rpm.

Use/abuse to taste.

```
# MS spec file for php-pecl-amqp
# remirepo spec file for php-pecl-amqp
# with SCL compatibility, from:
#
# Fedora spec file for php-pecl-amqp
#
# Copyright (c) 2012-2016 Remi Collet
# License: CC-BY-SA
# http://creativecommons.org/licenses/by-sa/4.0/
#
# Please, preserve the changelog entries
#
%{!?php_inidir:  %global php_inidir  %{_sysconfdir}/php.d}
%{!?__pecl:      %global __pecl      %{_bindir}/pecl}
%{!?__php:       %global __php       %{_bindir}/php}

%global with_zts    0%{?__ztsphp:1}
%global with_tests  0%{?_with_tests:1}
%global pecl_name   amqp
%if "%{php_version}" < "5.6"
%global ini_name    %{pecl_name}.ini
%else
%global ini_name    40-%{pecl_name}.ini
%endif
%global php_base    php70u
#global prever      beta4

Summary:       Communicate with any AMQP compliant server
Name:          %{php_base}-pecl-%{pecl_name}
Version:       1.7.0alpha2
Release:       1.ius%{dist}
License:       PHP
Group:         Development/Languages
URL:           http://pecl.php.net/package/amqp
Source0:       http://pecl.php.net/get/%{pecl_name}-%{version}%{?prever}.tgz


BuildRequires: %{php_base}-devel
BuildRequires: %{php_base}-pear
BuildRequires: librabbitmq-devel >= 0.5.0

Requires:         %{php_base}(zend-abi) = %{php_zend_api}
Requires:         %{php_base}(api) = %{php_core_api}

Requires(post):   %{__pecl}
Requires(postun): %{__pecl}

Provides:         php-%{pecl_name} = %{version}
Provides:         php-%{pecl_name}%{?_isa} = %{version}
Provides:         php-pecl(%{pecl_name}) = %{version}
Provides:         php-pecl(%{pecl_name})%{?_isa} = %{version}

Provides:         %{php_base}-%{pecl_name} = %{version}
Provides:         %{php_base}-%{pecl_name}%{?_isa} = %{version}
Provides:         %{php_base}(%{pecl_name}) = %{version}
Provides:         %{php_base}(%{pecl_name})%{?_isa} = %{version}

%if 0%{?fedora} < 20 && 0%{?rhel} < 7
# filter private shared
%{?filter_provides_in: %filter_provides_in %{_libdir}/.*\.so$}
%{?filter_setup}
%endif

%description
This extension can communicate with any AMQP spec 0-9-1 compatible server,
such as RabbitMQ, OpenAMQP and Qpid, giving you the ability to create and
delete exchanges and queues, as well as publish to any exchange and consume
from any queue.


%prep
%setup -q -c

# Don't install/register tests
sed -e 's/role="test"/role="src"/' -i package.xml

mv %{pecl_name}-%{version}%{?prever} NTS
cd NTS

# Upstream often forget to change this
extver=$(sed -n '/#define PHP_AMQP_VERSION/{s/.* "//;s/".*$//;p}' php_amqp.h)
if test "x${extver}" != "x%{version}%{?prever}"; then
   : Error: Upstream version is ${extver}, expecting %{version}%{?prever}.
   exit 1
fi
cd ..

cat > %{ini_name} << 'EOF'
; Enable %{pecl_name} extension module
extension = %{pecl_name}.so

; Whether calls to AMQPQueue::get() and AMQPQueue::consume()
; should require that the client explicitly acknowledge messages.
; Setting this value to 1 will pass in the AMQP_AUTOACK flag to
; the above method calls if the flags field is omitted.
;amqp.auto_ack = 0

; The host to which to connect.
;amqp.host = localhost

; The login to use while connecting to the broker.
;amqp.login = guest

; The password to use while connecting to the broker.
;amqp.password = guest

; The port on which to connect.
;amqp.port = 5672

; The number of messages to prefect from the server during a
; call to AMQPQueue::get() or AMQPQueue::consume() during which
; the AMQP_AUTOACK flag is not set.
;amqp.prefetch_count = 3

; The virtual host on the broker to which to connect.
;amqp.vhost = /

; Timeout
;amqp.timeout =
;amqp.read_timeout = 0
;amqp.write_timeout = 0
;amqp.connect_timeout = 0

;amqp.channel_max = 256
;amqp.frame_max = 131072
;amqp.heartbeat = 0
EOF

%if %{with_zts}
cp -pr NTS ZTS
%endif


%build
cd NTS
%{_bindir}/phpize
%configure --with-php-config=%{_bindir}/php-config
make %{?_smp_mflags}

%if %{with_zts}
cd ../ZTS
%{_bindir}/zts-phpize
%configure --with-php-config=%{_bindir}/zts-php-config
make %{?_smp_mflags}
%endif


%install
make -C NTS install INSTALL_ROOT=%{buildroot}

# Drop in the bit of configuration
install -Dpm 644 %{ini_name} %{buildroot}%{php_inidir}/%{ini_name}

# Install XML package description
install -Dpm 644 package.xml %{buildroot}%{pecl_xmldir}/%{name}.xml

%if %{with_zts}
make -C ZTS install INSTALL_ROOT=%{buildroot}
install -Dpm 644 %{ini_name} %{buildroot}%{php_ztsinidir}/%{ini_name}
%endif

# Documentation
cd NTS
for i in $(grep 'role="doc"' ../package.xml | sed -e 's/^.*name="//;s/".*$//')
do install -Dpm 644 $i %{buildroot}%{pecl_docdir}/%{pecl_name}/$i
done


%check
: Minimal load test for NTS extension
%{__php} --no-php-ini \
    --define extension=NTS/modules/%{pecl_name}.so \
    -m | grep %{pecl_name}

%if %{with_zts}
: Minimal load test for ZTS extension
%{__ztsphp} --no-php-ini \
    --define extension=ZTS/modules/%{pecl_name}.so \
    -m | grep %{pecl_name}
%endif

%if %{with_tests}
mkdir log run base
: Launch the RabbitMQ service
export RABBITMQ_PID_FILE=$PWD/run/pid
export RABBITMQ_LOG_BASE=$PWD/log
export RABBITMQ_MNESIA_BASE=$PWD/base
/usr/lib/rabbitmq/bin/rabbitmq-server &>log/output &
/usr/lib/rabbitmq/bin/rabbitmqctl wait $RABBITMQ_PID_FILE

ret=0
pushd NTS
: Run the upstream test Suite for NTS extension
TEST_PHP_EXECUTABLE=%{__php} \
TEST_PHP_ARGS="-n -d extension=$PWD/modules/%{pecl_name}.so" \
NO_INTERACTION=1 \
REPORT_EXIT_STATUS=1 \
%{__php} -n run-tests.php --show-diff || ret=1
popd

%if %{with_zts}
pushd ZTS
: Run the upstream test Suite for ZTS extension
TEST_PHP_EXECUTABLE=%{__ztsphp} \
TEST_PHP_ARGS="-n -d extension=$PWD/modules/%{pecl_name}.so" \
NO_INTERACTION=1 \
REPORT_EXIT_STATUS=1 \
%{__ztsphp} -n run-tests.php --show-diff || ret=1
popd
%endif

: Cleanup
if [ -s $RABBITMQ_PID_FILE ]; then
   kill $(cat $RABBITMQ_PID_FILE)
fi
rm -rf log run base

exit $ret
%endif


%post
%{pecl_install} %{pecl_xmldir}/%{name}.xml >/dev/null || :


%postun
if [ $1 -eq 0 ] ; then
    %{pecl_uninstall} %{pecl_name} >/dev/null || :
fi


%files
%defattr(-,root,root,-)
%doc %{pecl_docdir}/%{pecl_name}
%{pecl_xmldir}/%{name}.xml

%config(noreplace) %{php_inidir}/%{ini_name}
%{php_extdir}/%{pecl_name}.so

%if %{with_zts}
%config(noreplace) %{php_ztsinidir}/%{ini_name}
%{php_ztsextdir}/%{pecl_name}.so
%endif


%changelog
* Wed Mar 2 2016 Mark Steele <mark@control-alt-del.orG> - 1.6.1-2
- Build to follow standard paths

* Wed Nov 25 2015 Remi Collet <remi@fedoraproject.org> - 1.6.1-1
- update to 1.6.1 (stable)

* Tue Nov  3 2015 Remi Collet <remi@fedoraproject.org> - 1.6.0-2
- update to 1.6.0 (stable)
- fix typo in config file

* Fri Sep 18 2015 Remi Collet <remi@fedoraproject.org> - 1.6.0-0.4.beta4
- open https://github.com/pdezwart/php-amqp/pull/178 - librabbitmq 0.5
- open https://github.com/pdezwart/php-amqp/pull/179 --with-libdir

* Fri Sep 18 2015 Remi Collet <remi@fedoraproject.org> - 1.6.0-0.3.beta4
- update to 1.6.0beta4

* Fri Jun 19 2015 Remi Collet <remi@fedoraproject.org> - 1.6.0-0.2.beta3
- allow build against rh-php56 (as more-php56)

* Mon Apr 20 2015 Remi Collet <remi@fedoraproject.org> - 1.6.0-0.1.beta3
- update to 1.6.0beta3
- drop runtime dependency on pear, new scriptlets
- don't install/register tests

* Wed Dec 24 2014 Remi Collet <remi@fedoraproject.org> - 1.4.0-1.1
- Fedora 21 SCL mass rebuild

* Mon Aug 25 2014 Remi Collet <rcollet@redhat.com> - 1.4.0-1
- improve SCL build

* Tue Apr 15 2014 Remi Collet <remi@fedoraproject.org> - 1.4.0-1
- update to 1.6.0 (stable)

* Wed Apr  9 2014 Remi Collet <remi@fedoraproject.org> - 1.4.0-0.3.beta2
- add numerical prefix to extension configuration file

* Sun Mar  9 2014 Remi Collet <remi@fedoraproject.org> - 1.4.0-0.2.beta2
- update to 1.4.0beta2

* Thu Jan 16 2014 Remi Collet <remi@fedoraproject.org> - 1.4.0-0.1.beta1
- update to 1.4.0beta1
- adapt for SCL

* Mon Nov 25 2013 Remi Collet <remi@fedoraproject.org> - 1.3.0-1
- update to 1.3.0 (beta)
- cleanups for Copr
- install doc in pecl doc_dir
- install tests in pecl test_dir (in devel)
- add --with tests option to run upstream tests during build

* Sat Sep 28 2013 Remi Collet <remi@fedoraproject.org> - 1.2.0-2
- rebuild with librabbitmq 0.4.1

* Thu May 30 2013 Remi Collet <remi@fedoraproject.org> - 1.2.0-1
- Update to 1.2.0 (stable)

* Fri Apr 19 2013 Remi Collet <remi@fedoraproject.org> - 1.0.10-1
- Update to 1.0.10 (stable)

* Wed Mar 13 2013 Remi Collet <remi@fedoraproject.org> - 1.0.9-3
- rebuild for new librabbitmq

* Tue Nov 13 2012 Remi Collet <remi@fedoraproject.org> - 1.0.9-1
- update to 1.0.9 (stable)

* Mon Nov 12 2012 Remi Collet <remi@fedoraproject.org> - 1.0.8-1
- update to 1.0.8 (stable)
- build ZTS extension
- also provides php-amqp

* Wed Sep 12 2012 Remi Collet <remi@fedoraproject.org> - 1.0.7-1
- update to 1.0.7 (stable)
- cleanups

* Mon Aug 27 2012 Remi Collet <remi@fedoraproject.org> - 1.0.5-1
- update to 1.0.5
- LICENSE now provided in upstream tarball

* Wed Aug 01 2012 Remi Collet <remi@fedoraproject.org> - 1.0.4-1
- update to 1.0.4

* Sat Jul 21 2012 Fedora Release Engineering <rel-eng@lists.fedoraproject.org> - 1.0.3-2
- Rebuilt for https://fedoraproject.org/wiki/Fedora_18_Mass_Rebuild

* Sat May 19 2012 Remi Collet <remi@fedoraproject.org> - 1.0.3-1
- update to 1.0.3
- add extension version check (and fix)

* Mon Mar 19 2012 Remi Collet <remi@fedoraproject.org> - 1.0.1-3
- clean EL-5 stuff as requires php 5.2.0, https://bugs.php.net/61351

* Sat Mar 10 2012 Remi Collet <remi@fedoraproject.org> - 1.0.1-2
- rebuild for PHP 5.4

* Sat Mar 10 2012 Remi Collet <remi@fedoraproject.org> - 1.0.1-1
- Initial RPM release without ZTS extension
- open request for LICENSE file https://bugs.php.net/61337
```

**Comment by carlwgeorge** [comment]
Whoops, I misread the changelog initially.  There isn't a stable upstream release with PHP 7 compatibility.  IUS stands for "Inline with Upstream Stable", and we have a strong preference on sticking with the stable upstream versions of packages.  PHP 7 compatibility appears to still be in an alpha status upstream.  We'll put this request on hold until 1.7.0 stable is released.

**Comment by carlwgeorge** [comment]
Upstream has released a stable version (1.7.0) that is compatible with PHP7.  The IUS spec file will be maintained [here](https://github.com/iuscommunity-pkg/php70u-pecl-amqp).  Builds are in progress.

**Comment by carlwgeorge** [comment]
php70u-pecl-amqp has been completed and will move to the testing repos during tonight's automation.  Let us know how it works out for you.

**Comment by carlwgeorge** [comment]
@marksteele Have you had a chance to test the packages?  Once you confirm that they work as expected we will promote them to the stable repositories.

**Comment by guilhermeblanco** [comment]
@carlwgeorge I work with @marksteele at e-Block. Just finished testing and package seems to be working as intended. Thanks! =)

**Comment by b-harper** [comment]
@guilhermeblanco,

Thanks for the feedback.  We will push these packages to the stable repositories during tonight's automation run.

**Comment by guilhermeblanco** [comment]
@b-harper @carlwgeorge keep in mind I'd strongly suggest this to be moved together with librabbitmq (https://github.com/iuscommunity/wishlist/issues/53).

**Comment by b-harper** [comment]
@guilhermeblanco 

Please note that these php70u-pecl-amqp packages were built with stock librabbitmq-devel-0.5.2-1 packages and no IUS librabbitmq packages have been built.

**Comment by guilhermeblanco** [comment]
@b-harper my previous build (the one we opened the request) was built against 0.7.1... so I think I'll experience the same segfaults I did when I tried with 0.5.2... =( I've monitored with a debug build for weeks trying to drill down the problem on pecl-amqp, without any success (basically with debug build it never happened).
This means that I might potentially get the segfaults I did again... =(

**Comment by b-harper** [comment]
Besides the following line from @marksteele spec file, this is the first time anyone has mentioned anything about librabbitmq versioning:

```
BuildRequires: librabbitmq-devel >= 0.5.0
```

Can you confirm which version of librabbitmq you used for testing?  Could we hold off on promoting the current php70u-pecl-amqp packages to stable for now?

**Comment by guilhermeblanco** [comment]
I have `http://rpms.famillecollet.com/enterprise/7/remi/x86_64/librabbitmq-last-0.7.1-1.el7.remi.x86_64.rpm` installed when used your package for testing. 
I'll uninstall 0.7.1, install 0.5.2 and testing pecl-amqp... should get back to you in a day or two.

**Comment by carlwgeorge** [comment]
Our packages link against `librabbitmq.so.1`.  Even if you installed Remi's librabbitmq-last 0.7.1 package (a parallel installation that provides `librabbitmq.so.4`), our amqp package was still using to older 0.5.2 library.

```
$ rpm -qpl --provides librabbitmq-last-0.7.1-1.el7.remi.x86_64.rpm | grep 'librabbitmq.so.*()'
librabbitmq.so.4()(64bit)
$ repoquery --enablerepo ius-testing --requires php70u-pecl-amqp | grep 'librabbitmq.so.*()'
librabbitmq.so.1()(64bit)
```

We'll hold off on promoting this package to the stable repositories until you have completed your additional testing.

**Comment by carlwgeorge** [comment]
Have yall had a chance to test this package further?  The test suite runs successfully, [except for one test](https://github.com/pdezwart/php-amqp/issues/232) that requires a newer version of rabbitmq-server to talk to.

**Comment by carlwgeorge** [comment]
@marksteele @guilhermeblanco Any update on testing this?

**Comment by guilhermeblanco** [comment]
@carlwgeorge I'm sorry, I thought I had answered you already.
Yes, I did have the ability to test this further and it works properly with librabbitmq 0.5.2.
It's good to go... =)

**Comment by carlwgeorge** [comment]
Sounds good.  The package will be promoted to the stable repositories during tonight's automation.  Thanks for participating in IUS!

**Comment by guilhermeblanco** [comment]
@carlwgeorge you're welcome! =)

**Comment by carlwgeorge** [comment]
FYI, since we first created this package, upstream has released the 1.7.1 update.  We are building that now, and it will show up in our testing repos soon.  We only require feedback for the initial package, so the 1.7.1 update will get automatically promoted to the stable repos after two weeks.  But it will be there for you if you would like to test early.
